Dev tests on Debian 9 : fuse crash when mountpoint not created

Running
`py.test --log-level=DEBUG --durations=10 -v tests --log-level=DEBUG --durations=10 -v--runmountpoint --runslow -n auto --max-worker-restart=0 -x`
on Debian 9
with Python 3.7.8 installed from source
running in a venv

```
[gw3] [ 91%] FAILED tests/core/mountpoint/test_base.py::test_base_mountpoint_not_created

======================================================= FAILURES ========================================================
___________________________________________ test_base_mountpoint_not_created ____________________________________________
[gw3] linux -- Python 3.7.8 /home/test/python37-parsectests/bin/python3.7

    def _run_fuse_thread():
        logger.info("Starting fuse thread...", mountpoint=mountpoint_path)
        try:
            fuse_thread_started.set()
            FUSE(
                fuse_operations,
                str(mountpoint_path.absolute()),
                foreground=True,
                auto_unmount=True,
                encoding=encoding,
>               **config,
            )

../../../python37-parsectests/lib/python3.7/site-packages/parsec/core/mountpoint/fuse_runner.py:139:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <fuse.FUSE object at 0x7f6912dd6890>
operations = <parsec.core.mountpoint.fuse_operations.FuseOperations object at 0x7f6912dd6bd0>
mountpoint = '/tmp/pytest-of-test/pytest-0/popen-gw3/test_base_mountpoint_not_creat0/base_mountpoint/dummy/dummy/dummy/w'
raw_fi = False, encoding = 'utf-8', kwargs = {'auto_unmount': True, 'fsname': 'FuseOperations'}
args = [b'fuse', b'-f', b'-o', b'auto_unmount,fsname=FuseOperations', b'/tmp/pytest-of-test/pytest-0/popen-gw3/test_base_mountpoint_not_creat0/base_mountpoint/dummy/dummy/dummy/w']
argv = <fuse.c_char_p_Array_5 object at 0x7f6913e93440>, fuse_ops = <fuse.fuse_operations object at 0x7f6913e937a0>
ent = ('ioctl', <class 'ctypes.CFUNCTYPE.<locals>.CFunctionType'>), name = 'ioctl'
prototype = <class 'ctypes.CFUNCTYPE.<locals>.CFunctionType'>, check_name = 'ioctl'

    def __init__(self, operations, mountpoint, raw_fi=False, encoding='utf-8',
                 **kwargs):

        '''
        Setting raw_fi to True will cause FUSE to pass the fuse_file_info
        class as is to Operations, instead of just the fh field.

        This gives you access to direct_io, keep_cache, etc.
        '''

        self.operations = operations
        self.raw_fi = raw_fi
        self.encoding = encoding
        self.__critical_exception = None

        self.use_ns = getattr(operations, 'use_ns', False)
        if not self.use_ns:
            warnings.warn(
                'Time as floating point seconds for utimens is deprecated!\n'
                'To enable time as nanoseconds set the property "use_ns" to '
                'True in your operations class or set your fusepy '
                'requirements to <4.',
                DeprecationWarning)

        args = ['fuse']

        args.extend(flag for arg, flag in self.OPTIONS
                    if kwargs.pop(arg, False))

        kwargs.setdefault('fsname', operations.__class__.__name__)
        args.append('-o')
        args.append(','.join(self._normalize_fuse_options(**kwargs)))
        args.append(mountpoint)

        args = [arg.encode(encoding) for arg in args]
        argv = (ctypes.c_char_p * len(args))(*args)

        fuse_ops = fuse_operations()
        for ent in fuse_operations._fields_:
            name, prototype = ent[:2]

            check_name = name

            # ftruncate()/fgetattr() are implemented in terms of their
            # non-f-prefixed versions in the operations object
            if check_name in ["ftruncate", "fgetattr"]:
                check_name = check_name[1:]

            val = getattr(operations, check_name, None)
            if val is None:
                continue

            # Function pointer members are tested for using the
            # getattr(operations, name) above but are dynamically
            # invoked using self.operations(name)
            if hasattr(prototype, 'argtypes'):
                val = prototype(partial(self._wrapper, getattr(self, name)))

            setattr(fuse_ops, name, val)

        try:
            old_handler = signal(SIGINT, SIG_DFL)
        except ValueError:
            old_handler = SIG_DFL

        err = _libfuse.fuse_main_real(
            len(args), argv, ctypes.pointer(fuse_ops),
            ctypes.sizeof(fuse_ops),
            None)

        try:
            signal(SIGINT, old_handler)
        except ValueError:
            pass

        del self.operations     # Invoke the destructor
        if self.__critical_exception:
            raise self.__critical_exception
        if err:
>           raise RuntimeError(err)
E           RuntimeError: 1

../../../python37-parsectests/lib/python3.7/site-packages/fuse.py:711: RuntimeError

The above exception was the direct cause of the following exception:

value = <trio.Nursery object at 0x7f691943f2d0>

    async def yield_(value=None):
>       return await _yield_(value)

../../../python37-parsectests/lib/python3.7/site-packages/async_generator/_impl.py:106:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../../python37-parsectests/lib/python3.7/site-packages/async_generator/_impl.py:99: in _yield_
    return (yield _wrap(value))
tests/core/mountpoint/test_base.py:76: in test_base_mountpoint_not_created
    await mountpoint_manager.mount_workspace(wid)
../../../python37-parsectests/lib/python3.7/site-packages/parsec/core/mountpoint/manager.py:184: in mount_workspace
    runner_task = await self._mount_workspace_helper(workspace)
../../../python37-parsectests/lib/python3.7/site-packages/parsec/core/mountpoint/manager.py:158: in _mount_workspace_helper
    runner_task = await start_task(self._nursery, curried_runner)
../../../python37-parsectests/lib/python3.7/site-packages/parsec/utils.py:102: in start_task
    return await nursery.start(TaskStatus.wrap_task, corofn, *args, name=name)
../../../python37-parsectests/lib/python3.7/site-packages/parsec/service_nursery.py:183: in start
    return await type(nursery).start(nursery, wrap_child, name=name or async_fn)
../../../python37-parsectests/lib/python3.7/site-packages/trio/_core/_run.py:952: in start
    old_nursery.start_soon(thunk, *args, name=name)
../../../python37-parsectests/lib/python3.7/site-packages/trio/_core/_run.py:730: in __aexit__
    raise combined_error_from_nursery
../../../python37-parsectests/lib/python3.7/site-packages/parsec/service_nursery.py:181: in wrap_child
    await async_fn(*args, task_status=task_status)
../../../python37-parsectests/lib/python3.7/site-packages/parsec/utils.py:89: in wrap_task
    value = await nursery.start(corofn, *args)
../../../python37-parsectests/lib/python3.7/site-packages/parsec/service_nursery.py:183: in start
    return await type(nursery).start(nursery, wrap_child, name=name or async_fn)
../../../python37-parsectests/lib/python3.7/site-packages/trio/_core/_run.py:952: in start
    old_nursery.start_soon(thunk, *args, name=name)
../../../python37-parsectests/lib/python3.7/site-packages/trio/_core/_run.py:730: in __aexit__
    raise combined_error_from_nursery
../../../python37-parsectests/lib/python3.7/site-packages/parsec/service_nursery.py:181: in wrap_child
    await async_fn(*args, task_status=task_status)
../../../python37-parsectests/lib/python3.7/site-packages/parsec/core/mountpoint/manager.py:153: in curried_runner
    task_status=task_status,
../../../python37-parsectests/lib/python3.7/site-packages/parsec/core/mountpoint/fuse_runner.py:167: in fuse_mountpoint_runner
    task_status.started(mountpoint_path)
../../../python37-parsectests/lib/python3.7/site-packages/async_generator/_util.py:53: in __aexit__
    await self._agen.athrow(type, value, traceback)
../../../python37-parsectests/lib/python3.7/site-packages/parsec/utils.py:182: in open_service_nursery
    yield nursery
../../../python37-parsectests/lib/python3.7/site-packages/async_generator/_util.py:53: in __aexit__
    await self._agen.athrow(type, value, traceback)
../../../python37-parsectests/lib/python3.7/site-packages/parsec/service_nursery.py:191: in open_service_nursery
    child_task_scopes.shield = False
../../../python37-parsectests/lib/python3.7/site-packages/trio/_core/_run.py:730: in __aexit__
    raise combined_error_from_nursery
../../../python37-parsectests/lib/python3.7/site-packages/parsec/service_nursery.py:158: in wrap_child
    await coro
../../../python37-parsectests/lib/python3.7/site-packages/trio/_threads.py:310: in to_thread_run_sync
    return await trio.hazmat.wait_task_rescheduled(abort)
../../../python37-parsectests/lib/python3.7/site-packages/trio/_core/_traps.py:165: in wait_task_rescheduled
    return (await _async_yield(WaitTaskRescheduled(abort_func))).unwrap()
../../../python37-parsectests/lib/python3.7/site-packages/outcome/_sync.py:111: in unwrap
    raise captured_error
../../../python37-parsectests/lib/python3.7/site-packages/trio/_threads.py:262: in do_release_then_return_result
    return result.unwrap()
../../../python37-parsectests/lib/python3.7/site-packages/outcome/_sync.py:111: in unwrap
    raise captured_error
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

    def _run_fuse_thread():
        logger.info("Starting fuse thread...", mountpoint=mountpoint_path)
        try:
            fuse_thread_started.set()
            FUSE(
                fuse_operations,
                str(mountpoint_path.absolute()),
                foreground=True,
                auto_unmount=True,
                encoding=encoding,
                **config,
            )

        except Exception as exc:
            try:
                errcode = errno.errorcode[exc.args[0]]
            except (KeyError, IndexError):
                errcode = f"Unknown error code: {exc}"
            raise MountpointDriverCrash(
                f"Fuse has crashed on {mountpoint_path}: {errcode}"
>           ) from exc
E           parsec.core.mountpoint.exceptions.MountpointDriverCrash: Fuse has crashed on /tmp/pytest-of-test/pytest-0/popen-gw3/test_base_mountpoint_not_creat0/base_mountpoint/dummy/dummy/dummy/w: EPERM

../../../python37-parsectests/lib/python3.7/site-packages/parsec/core/mountpoint/fuse_runner.py:149: MountpointDriverCrash
------------------------------------------------- Captured stderr call --------------------------------------------------
fuse: failed to exec fusermount: No such file or directory
--------------------------------------------------- Captured log call ---------------------------------------------------
DEBUG    parsec.event_bus:event_bus.py:54 2020-07-08 14:07:04 [debug    ] Send event                     [parsec.event_bus] event_type=<CoreEvent.FS_ENTRY_UPDATED: 'fs.entry.updated'> id=EntryID('6897c693-18bd-4004-a539-a63afd4481ac')
DEBUG    parsec.event_bus:event_bus.py:54 2020-07-08 14:07:04 [debug    ] Send event                     [parsec.event_bus] event_type=<CoreEvent.FS_WORKSPACE_CREATED: 'fs.workspace.created'> new_entry=WorkspaceEntry(name='w', id=EntryID('481384a4-d7e8-4624-97c1-84bb5ae7d4a9'), key=<parsec.crypto.SecretKey object at 0x7f6919593530>, encryption_revision=1, encrypted_on=<Pendulum [2020-07-08T14:07:04.836744+00:00]>, role_cached_on=<Pendulum [2020-07-08T14:07:04.836744+00:00]>, role=<RealmRole.OWNER: 'OWNER'>)
DEBUG    parsec.event_bus:event_bus.py:54 2020-07-08 14:07:04 [debug    ] Send event                     [parsec.event_bus] event_type=<CoreEvent.FS_ENTRY_UPDATED: 'fs.entry.updated'> id=EntryID('481384a4-d7e8-4624-97c1-84bb5ae7d4a9') workspace_id=EntryID('481384a4-d7e8-4624-97c1-84bb5ae7d4a9')
DEBUG    parsec.event_bus:event_bus.py:54 2020-07-08 14:07:04 [debug    ] Send event                     [parsec.event_bus] event_type=<CoreEvent.FS_ENTRY_UPDATED: 'fs.entry.updated'> id=EntryID('64035b6b-8da1-43fe-9150-b80ecc46b742') workspace_id=EntryID('481384a4-d7e8-4624-97c1-84bb5ae7d4a9')
DEBUG    parsec.event_bus:event_bus.py:54 2020-07-08 14:07:04 [debug    ] Send event                     [parsec.event_bus] event_type=<CoreEvent.MOUNTPOINT_STARTING: 'mountpoint.starting'> mountpoint=PosixPath('/tmp/pytest-of-test/pytest-0/popen-gw3/test_base_mountpoint_not_creat0/base_mountpoint/dummy/dummy/dummy/w') workspace_id=EntryID('481384a4-d7e8-4624-97c1-84bb5ae7d4a9')
INFO     parsec.core.mountpoint.fuse_runner:fuse_runner.py:130 2020-07-08 14:07:04 [info     ] Starting fuse thread...        [parsec.core.mountpoint.fuse_runner] mountpoint=PosixPath('/tmp/pytest-of-test/pytest-0/popen-gw3/test_base_mountpoint_not_creat0/base_mountpoint/dummy/dummy/dummy/w')
DEBUG    parsec.event_bus:event_bus.py:54 2020-07-08 14:07:04 [debug    ] Send event                     [parsec.event_bus] event_type=<CoreEvent.MOUNTPOINT_STOPPED: 'mountpoint.stopped'> mountpoint=PosixPath('/tmp/pytest-of-test/pytest-0/popen-gw3/test_base_mountpoint_not_creat0/base_mountpoint/dummy/dummy/dummy/w') workspace_id=EntryID('481384a4-d7e8-4624-97c1-84bb5ae7d4a9')
=============================================== slowest 10 test durations ===============================================
18.00s call     tests/backend/realm/test_shuffle_roles.py::test_shuffle_roles
17.27s call     tests/test_cli.py::test_full_run[no_ssl]
16.99s call     tests/core/fs/test_fs_idempotent_sync.py::test_fs_online_idempotent_sync
16.66s call     tests/test_cli.py::test_full_run[ssl]
14.01s call     tests/test_cli.py::test_apiv1_full_run[ssl]
13.77s call     tests/test_cli.py::test_apiv1_full_run[no_ssl]
11.84s call     tests/core/fs/workspacefs/test_file_transactions.py::test_file_operations
8.59s call     tests/core/fs/test_fs_offline_restart_and_rwfile.py::test_fs_offline_restart_and_rwfile
7.54s call     tests/core/fs/test_fs_offline_restart_and_tree.py::test_fs_offline_restart_and_tree
6.35s call     tests/core/fs/workspacefs/test_entry_transactions.py::test_entry_transactions
================================================ short test summary info ================================================
FAILED tests/core/mountpoint/test_base.py::test_base_mountpoint_not_created - parsec.core.mountpoint.exceptions.Mountp...
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! xdist.dsession.Interrupted: stopping after 1 failures !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
=========================== 1 failed, 1092 passed, 105 skipped, 3 xfailed in 80.13s (0:01:20) ===========================
```

Is it an issue with an outdated fuse driver? If so, how to update it?

Ideally, mountpoint/fuse_runner.py should catch this crash when mountpoint was not created, or even better (if possible and easy) prevent a mounting attempt if there's no mountpoint (means add a mountpoint detection before mounting).

--- a/.azure-pipelines.yml
+++ b/.azure-pipelines.yml
@@ -360,9 +360,9 @@ jobs:
 - job: 'linux_snap_test'
   displayName: 'Never trust a snap'
   dependsOn: linux_snap_build
-  timeoutInMinutes: 5
+  timeoutInMinutes: 10
   pool:
-    vmImage: 'ubuntu-18.04'  # Shame, snap doesn't support docker...
+    vmImage: 'ubuntu-latest'
   steps:
   # Note there no point specifying Python version given snap doesn't take it into account
   - checkout: none
@@ -374,17 +374,38 @@ jobs:
       # sudo apt-get install -y xvfb xorg
       sudo snap install --classic --dangerous $(Pipeline.Workspace)/linux_amd64_snap/parsec*.snap
       parsec.cli --version
+      # Make sure snap contains core but not backend
+      NOT_AVAILABLES=$(parsec.cli --help | grep -i "not available")
+      test "$(echo $NOT_AVAILABLES | grep -i 'backend')"
+      test -z "$(echo $NOT_AVAILABLES | grep -i 'core')"
       # xvfb-run parsec &
       # sleep 5
       # jobs  # returns error if parsec is not started
       # kill %1
     displayName: 'Test Snap'
+
+
+#################################################################################
+
+
+- job: 'linux_snap_upload'
+  displayName: 'Upload snap'
+  dependsOn: linux_snap_test
+  timeoutInMinutes: 10
+  pool:
+    vmImage: 'ubuntu-latest'
+  steps:
+  - checkout: none
+  - download: current
+    artifact: linux_amd64_snap
   - bash: |
       sudo snap install snapcraft --classic
       echo $SNAPCRAFT_CREDENTIALS | snapcraft login --with -
-      snapcraft push --release=edge $(Pipeline.Workspace)/linux_amd64_snap/parsec*.snap
+    displayName: 'Bootstrap'
     env:
       SNAPCRAFT_CREDENTIALS: $(SNAPCRAFT_CREDENTIALS)
+  - bash: |
+      snapcraft upload --release=edge $(Pipeline.Workspace)/linux_amd64_snap/parsec*.snap
     displayName: 'Upload Snap'
 
 

--- a/newsfragments/1296.misc.rst
+++ b/newsfragments/1296.misc.rst
@@ -1,0 +1,1 @@
+Fix mount error when using Snap package on Debian when fuse is not installed.

--- a/packaging/snap/snap/snapcraft.yaml
+++ b/packaging/snap/snap/snapcraft.yaml
@@ -42,6 +42,9 @@ parts:
       # Ugly hack given snapcraft doesn't support extra requirements...
       sed -i "s#install_requires=requirements#install_requires=requirements + extra_requirements['core']#g" setup.py
     stage-packages:
+      # fuse package contains fusermount which is sometime needed by libfuse2
+      # (see: https://github.com/Scille/parsec-cloud/issues/1253)
+      - fuse
       - libfuse2
       - libxext6
       - libx11-6

